### PR TITLE
Refactors configuration management

### PR DESCRIPTION
**Pull request by flovilmart**
### Motivation:

When adding a new startup option to parse-server, we had to update multiple files, including the CLI definitions, defaults, initializer etc... Which was a lot. This also lead to mismatches

This PR is intender to provide simple interfaces (backed by flow types) in order to generate the configuration definitions on the fly.

### resources/buildConfigDefinitions.js

It leverages a custom babel preset that transforms every entry from the interface into a definition. 
It also parses the comment in order to infer the help/env parameter
It writes the content of the file into src/Options/Definitions.js

### Config.js

- Changes from constructor to `Config.get(appId)` as it's more semantically correct
- Adds Config.put to abstract AppCache from ParseServer.js

### ParseServer.js

- Moves all controller creation logic into `Controllers/index.js`
- This makes the module more readable, and focused.

### Next Steps

- ~Get rid of the defaults, and use the executor to generate the default values~  💯 
- ~Reduce the code duplication between server and CLI.~ 💯

### CODE REVIEW

**Comment by montymxb** [comment]
I'm a bit off the grid at the moment. I'll be back behind a computer by
Monday, i can take a look at this then.

You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/parse-community/parse-server/pull/4271>, or mute the
thread

File: resources/buildConfigDefinitions.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+
+const plugin = function (babel) {
+  const t = babel.types;
+  const moduleExports = t.memberExpression(t.identifier('module'), t.identifier('exports'));
+  return {
+    visitor: {
+      Program: function(path) {
+        // Inject the parsers loader
```
Comment (acinader) [inline resources/buildConfigDefinitions.js]: parser's?

File: resources/buildConfigDefinitions.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        path.pushContainer("body", t.importDeclaration([t.importNamespaceSpecifier(t.identifier('parsers'))],
+          t.stringLiteral('./parsers')));
```
Comment (acinader) [inline resources/buildConfigDefinitions.js]: single quote on "body" -- "A foolish consistency is the hobgoblin of little minds" :).

File: src/defaults.js [hunk 3]
```diff
@@ -1,5 +1,4 @@
-import {nullParser} from './cli/utils/parsers';
-
+import {nullParser} from './Options/parsers';
```
Comment (acinader) [inline src/defaults.js]: add space between `{ null....`

**Comment by flovilmart** [comment]
@acinader yep my bad I thought I could get away not parsing the defaults, but I'Ll have to do it.

**Comment by flovilmart** [comment]
@acinader for the sake of simplicity, I decided to merge the system defaults with the provided defaults so all tests are OK, and the CLI still has it's defaults injected (mountPath, host, port)

**Comment by flovilmart** [comment]
@acinader could not resist the urge to push it further, now all default values are specified into the interface file, single place across the project!

**Comment by dplewis** [comment]
@flovilmart @acinader @montymxb I’m still new to open source projects how do you go about reviewing commits like this? What do you usually look for?

**Comment by flovilmart** [comment]
@dplewis 

usually, but not only.

- wether it makes sense as a feature
- potential breaking changes
- code smells, typos, etc...
- documentations / comments
- if everything looks clear
- if you're able to maintain the feature yourself (ie: you have a good understanding at what the code does) (ask for more comments in case, links etc..)
- more...

**Comment by dplewis** [comment]
@flovilmart Thanks for the knowledge!

**Comment by flovilmart** [comment]
The main goal of the PR is to drive the options from a single file. And use Babel to generate the definitions for the env vars, cli options and default values. The rest is additional refactors that makes everything a bit clearer (on the CLI module).
Next PR will be to refactor the ParseServer.js constructor so it’s easier to read.

**Comment by dplewis** [comment]
Nice, I started on a feature that I wanted to add to the CLI, this will make it easier.

File: resources/buildConfigDefinitions.js [hunk 4]
```diff
@@ -0,0 +1,221 @@
+function last(array) {
```
Comment (acinader) [inline resources/buildConfigDefinitions.js]: a comment for this file would be super helpful.  I'm reasonably familiar with the code base but could def use some clues as to what is going on here.....

File: resources/buildConfigDefinitions.js [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+  return array[array.length - 1];
+}
+
+const parsers = require('../src/Options/parsers');
```
Comment (acinader) [inline resources/buildConfigDefinitions.js]: requires at the top of file, please.

File: resources/buildConfigDefinitions.js [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+function toENV(key) {
+  var str = '';
```
Comment (acinader) [inline resources/buildConfigDefinitions.js]: prefer 'let' to 'var'

File: resources/buildConfigDefinitions.js [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+  let previousIsUpper = false;
+  for(let i = 0; i < key.length; i++) {
+    const char = key[i];
+    if (letters.indexOf(char) >= 0) {
+      if (!previousIsUpper) {
+        str += "_"
```
Comment (acinader) [inline resources/buildConfigDefinitions.js]: prefer `'`

File: resources/buildConfigDefinitions.js [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+const parsers = require('../src/Options/parsers');
+
+const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+function toENV(key) {
+  var str = '';
+  let previousIsUpper = false;
+  for(let i = 0; i < key.length; i++) {
+    const char = key[i];
+    if (letters.indexOf(char) >= 0) {
```
Comment (acinader) [inline resources/buildConfigDefinitions.js]: this should probably be a function that can be passed to map. (just thinking out loud. if it doesn't appeal to you, don't do it).

File: resources/buildConfigDefinitions.js [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    types: property.value.types,
+    typeAnnotation: property.value.typeAnnotation,
+    required: isRequired
+  };
+}
+
+
+function doInterface(iface) {
+  return iface.body.properties.map((prop) => {
+    return processProperty(prop, iface);
```
Comment (acinader) [inline resources/buildConfigDefinitions.js]: `prop => processProperty(prop, iface)`

File: resources/buildConfigDefinitions.js [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+  }).filter((e) => e !== undefined);
+}
+
+function mapperFor(elt, t) {
+  const p = t.identifier('parsers');
+  const wrap = function(identifier) {
+    return t.memberExpression(p, identifier);
```
Comment (acinader) [inline resources/buildConfigDefinitions.js]: `wrap = (identifier) => t.memberExpression(p, identifier);`

right?  if you prefer?

File: resources/buildConfigDefinitions.js [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+  }
+
```
Comment (acinader) [inline resources/buildConfigDefinitions.js]: maybe a switch?

File: resources/buildConfigDefinitions.js (hunk 4 above)
Comment (flovilmart) [reply resources/buildConfigDefinitions.js]: Yep for sure!

**Comment by flovilmart** [comment]
All great and valid comments, I was pretty sloppy on the builder :/

**Comment by montymxb** [comment]
I missed the entire review :/

**Comment by flovilmart** [comment]
You can still look at it, and I can tweak a few things if you lake, better late than never!

**Comment by montymxb** [comment]
Flipped through it and it looks nice! I like centralizing those configs, should make future management a bit easier.

**Comment by flovilmart** [comment]
yep, that was the goal, also exploring the babel compiler was something I wanted to do

**Pull request merged**